Lucid/Utils/Publisher+Extensions.swift

```diff
@@ -0,0 +1,340 @@
+//
+//  Publisher+Extensions.swift
+//  Lucid
+//
+//  Created by Stephane Magne on 1/28/22.
+//  Copyright © 2022 Scribd. All rights reserved.
+//
+
+#if !LUCID_REACTIVE_KIT
+import Combine
+
+// MARK: - Typealiases
+
+public typealias SafePublisher<Output> = AnyPublisher<Output, Never>
+
+// MARK: - Flat Map Errors
+
+public extension Publisher {
+
+    /// Transform an error into a result with a new error type. This allows the call-site to transform an error into successful output, or into a new error value.
+    /// - Note: The original Publisher will still be completed, so this can only be used on 'once' signals.
+    func flatMapError<F>(_ transform: @escaping (Failure) -> Result<Output, F>) -> AnyPublisher<Output, F> where F: Error {
+        return ErrorTransformSubject<Output, Failure, F>(observing: eraseToAnyPublisher(), transform: transform).eraseToAnyPublisher()
+    }
+
+    /// Transform an error into a new error type.
+    /// - Note: The original Publisher will still be completed, so this can only be used on 'once' signals.
+    func flatMapError<F>(_ transform: @escaping (Failure) -> F) -> AnyPublisher<Output, F> where F: Error {
+        return flatMapError { error -> Result<Output, F> in
+            return .failure(transform(error))
+        }
+    }
+
+    /// Transform an error into successful output.
+    /// - Note: The original Publisher will still be completed, so this can only be used on 'once' signals.
+    func flatMapError(_ transform: @escaping (Failure) -> Output) -> AnyPublisher<Output, Failure> {
+        return flatMapError { error -> Result<Output, Failure> in
+            return .success(transform(error))
+        }
+    }
+}
+
+// MARK: - Error Substitutions
+
+public extension AnyPublisher where Failure == ManagerError {
+
+    func substituteValueForNonCriticalFailure(_ value: Output) -> AnyPublisher<Output, ManagerError> {
+        return self
+            .substituteValueForInternetConnectionFailure(value)
+            .substituteValueForUserAccessFailure(value)
+    }
+
+    func substituteValueForInternetConnectionFailure(_ value: Output) -> AnyPublisher<Output, ManagerError> {
+
+        return flatMapError { managerError -> Result<Output, ManagerError> in
+            if managerError.isNetworkConnectionFailure {
+                return .success(value)
+            } else {
+                return .failure(managerError)
+            }
+        }
+    }
+
+    func substituteValueForUserAccessFailure(_ value: Output) -> AnyPublisher<Output, ManagerError> {
+        return flatMapError { managerError -> Result<Output, ManagerError> in
+            if managerError.isUserAccessFailure {
+                return .success(value)
+            } else {
+                return .failure(managerError)
+            }
+        }
+    }
+}
+
+// MARK: - Mutation Filtering
+
+public struct EntityMutationRule: OptionSet {
```

Comment:
Note: all of the code from here down to the start of `ErrorTransformSubject` is the same logic as in `Signal+Extensions`, but has transposed types (e.g. Element -> Output, Error -> Failure, Signal -> Publisher, etc)